### PR TITLE
Implement containerized-script execution method

### PR DESCRIPTION
**Pull request by Others**
This implements me and Henry's "cached containers" idea. Basically
we boot up containers before we need them so startup is faster.

This requires a bunch of threads creating containers in the
background. (See `idle_container_creator.rs`)

Unfortunately this isn't really faster than than the loading an
archive, which is unfortunate. I do see several opportunities
for optimization, however.

I've also refactored a lot of code, and added a lot of TODOs.

NB: There currently is a problem where containers can stick
around after the program terminates.

### CODE REVIEW

File: src/component/isolation.rs [hunk 1]
```diff
@@ -119,75 +117,33 @@ impl ProcessIsolationController for PythonUnsafeController {
...
-            .capture()?;
-        let load_exit_status = load_result.exit_status;
-        let load_stdout = String::from_utf8(load_result.stdout)?;
-        let load_stderr = String::from_utf8(load_result.stderr)?;
-
-        if !load_exit_status.success() {
-            return Err(
-                WorkerErrorKind::Docker(load_result.exit_status, load_stdout, load_stderr).into(),
-            );
+            return Err(WorkerErrorKind::UnsupportedPlatform("macos").into());
```
Comment (pcodes) [inline src/component/isolation.rs]: Solid solution

File: src/fs_utils.rs [hunk 2]
```diff
@@ -0,0 +1,18 @@
+use std::path::Path;
+
+use crate::error::{WorkerError, WorkerErrorKind};
+
+pub fn canonicalize(p: &Path) -> Result<String, WorkerError> {
```
Comment (pcodes) [inline src/fs_utils.rs]: Glad you added this, makes the calls to it much cleaner

File: src/docker/idle_container_creator.rs [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+use tempfile::TempDir;
+
+use crate::docker::call_docker_async;
+use crate::error::{WorkerError, WorkerErrorKind};
+use crate::fs_utils::get_filename;
+use crate::named_pipe::NamedPipe;
+
+// NOTE: the number of idle containers on the system is CONTAINER_CACHE_CHANNEL_SIZE + CACHE_POPULATOR_COUNT
+const CONTAINER_CACHE_CHANNEL_SIZE: usize = 3;
+const CACHE_POPULATOR_COUNT: usize = 2;
```
Comment (pcodes) [inline src/docker/idle_container_creator.rs]: Can we talk about what these numbers mean?

File: src/docker/idle_container_creator.rs [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+
+const CONTAINER_IMAGE_TAG: &str = "python";
+// 1000000000 seconds ~= 30 years
+const SLEEP_TIME: &str = "1000000000";
+
+const CONTAINER_PIPE_FOLDER: &str = "/home/io";
```
Comment (pcodes) [inline src/docker/idle_container_creator.rs]: I'm assuming there's a user called io?

File: src/docker/idle_container_creator.rs [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+}
+
+pub struct IdleContainerCreator {
+    cache_channel_receiver: Mutex<Receiver<Container>>,
+}
+
+impl IdleContainerCreator {
+    fn new() -> Self {
+        // Create the cache channel
+        let (sender, receiver) = sync_channel(CONTAINER_CACHE_CHANNEL_SIZE);
```
Comment (pcodes) [inline src/docker/idle_container_creator.rs]: I should probably look this up, but what is a cache channel?

File: src/component/isolation.rs [hunk 6]
```diff
@@ -209,6 +163,58 @@ impl ProcessIsolationController for DockerArchiveController {
 
 // TODO: Add a drop that gets rid of the loaded docker images after we're done with them
 
+#[derive(Debug)]
+pub struct ContainerizedScriptController {
+    executable_file: String,
+}
+
+impl ContainerizedScriptController {
```
Comment (pcodes) [inline src/component/isolation.rs]: What is this? What is it's purpose?

File: src/docker/idle_container_creator.rs (hunk 3 above)
Comment (Others) [reply src/docker/idle_container_creator.rs]: So `CONTAINER_CACHE_CHANNEL_SIZE` basically is how many containers can be cached in the channel between the cache populators and the user. The `CACHE_POPULATOR_COUNT` is how many threads are trying to populate that channel. Each thread may have started a container that isn't yet in the channel, so `CONTAINER_CACHE_CHANNEL_SIZE + CACHE_POPULATOR_COUNT = max number of cached containers`

File: src/docker/idle_container_creator.rs (hunk 4 above)
Comment (Others) [reply src/docker/idle_container_creator.rs]: Nope, I just wanted somewhere central

File: src/docker/idle_container_creator.rs (hunk 5 above)
Comment (Others) [reply src/docker/idle_container_creator.rs]: https://doc.rust-lang.org/std/sync/mpsc/fn.sync_channel.html

We just use it to store cached containers

File: src/component/isolation.rs (hunk 6 above)
Comment (Others) [reply src/component/isolation.rs]: I think I explained this abstraction to you previously, maybe look at the rest of `isolation.rs` to jog your memory

File: src/component/mod.rs [hunk 7]
```diff
@@ -146,7 +146,7 @@ impl ComponentManager {
         let memory_usage = self
             .system
             .memory()
-            .map(|mem| 1.0 - mem.total.as_u64() as f64 / mem.free.as_u64() as f64)
+            .map(|mem| 1.0 - mem.free.as_u64() as f64 / mem.total.as_u64() as f64)
```
Comment (pcodes) [inline src/component/mod.rs]: I should have caught this lol

File: src/docker/idle_container_creator.rs (hunk 4 above)
Comment (pcodes) [reply src/docker/idle_container_creator.rs]: Add a TODO here

File: src/docker/idle_container_creator.rs [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+    let desired_name = next_container_name();
+    let docker_subprocess = call_docker_async(&[
+        "run",
+        "-v",
+        &pipe_mount,
+        "--name",
+        &desired_name,
+        CONTAINER_IMAGE_TAG,
+        "sleep",
+        SLEEP_TIME,
```
Comment (pcodes) [inline src/docker/idle_container_creator.rs]: Just want this on record that this is super hacky

**Pull request merged**